ucrop/src/main/java/com/yalantis/ucrop/util/FileUtils.java

```diff
@@ -227,11 +232,42 @@ public static void copyFile(@NonNull String pathFrom, @NonNull String pathTo) th
             inputChannel = new FileInputStream(new File(pathFrom)).getChannel();
             outputChannel = new FileOutputStream(new File(pathTo)).getChannel();
             inputChannel.transferTo(0, inputChannel.size(), outputChannel);
-            inputChannel.close();
         } finally {
             if (inputChannel != null) inputChannel.close();
             if (outputChannel != null) outputChannel.close();
         }
     }
 
+    /**
+     * Copies one file into the other with the given Uris.
+     * In the event that the Uris are the same, trying to copy one file to the other
+     * will cause both files to become null.
+     * Simply skipping this step if the paths are identical.
+     *
+     * @param context The context from which to require the {@link android.content.ContentResolver}
+     * @param uriFrom Represents the source file
+     * @param uriTo Represents the destination file
+     */
+    public static void copyFile(@NonNull Context context, @NonNull Uri uriFrom, @NonNull Uri uriTo) throws IOException {
+        if (uriFrom.equals(uriTo)) {
+            return;
+        }
+
+        InputStream isFrom = null;
+        OutputStream osTo = null;
+        try {
+            isFrom = context.getContentResolver().openInputStream(uriFrom);
+            osTo = context.getContentResolver().openOutputStream(uriTo);
+
+            if (isFrom instanceof FileInputStream && osTo instanceof FileOutputStream) {
+                FileChannel inputChannel = ((FileInputStream)isFrom).getChannel();
+                FileChannel outputChannel = ((FileOutputStream)osTo).getChannel();
+                inputChannel.transferTo(0, inputChannel.size(), outputChannel);
+            }
```

Comment:
What if the statement won't be executed? Nothing will happen? Wouldn't be better to throw exception in `else` branch or at least log this case?

Reply:
Yeap, nothing will happen. The exif info will simply not be copied and you will end up with a destination file ok but with no exif info. If the rest of the code is right tough it should not happen. At least not in this part of the code. I've not thrown an exception to make less changes in the code since there are other places where this kind of thing happens too but I think we can change it. Throw an exception changes more code and can change more the actual behavior but either way can be done. Throw an exception or log.

Reply:
Forget what I've said about exif, I was thinking about another part of the code. This file copying method is used if there is no need to crop. It is replacing the method that does this based on file paths instead of *uris*. 
The mentioned statement is just a protection for the castings of `FileInputStream`. If some of the *uris* don't represent a file than the uCrop can't deal with those *uris" I think. There is nothing to do. But we surely should not eat that information! It may be indigestible!! :joy:
What do you think it is better? To trow an exception or to just log it? I don't know what would be a better exception here. IllegalArgumentException?

If I throw an exception here, it will be translated into an error for been displayed to the final user of the `UCropActivity`. If I just log the failure, some developer will have to dig into the log to know why the uCrop is not doing its work. What do you think it is the better approach for this case?

Reply:
In my opinion, it's better to throw an exception, cause it's more explicit way to inform developer or QA on development stage that input or output url isn't fittable. In this case the exception could be prevented and user won't see it. Or, at least, user will be informed that something went wrong and his file wasn't copied due to some error, so hi will be able to inform support about that.

I like your option to use `IllegalArgumentException` for this purpose.

Reply:
I agree. Let's do it this way.